Added support for WASM compilation

pmathew92 opened the pull request:
### 📋 Changes

This PR adds support for WASM compilation 

### 📎 References
#463 

### 🎯 Testing
This has been tested by running the example app with the  following command
`flutter run -d chrome --web-port 3000 --wasm `

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -52,29 +53,29 @@ class RedirectLoginOptions {
-  external String get version;
+extension type Auth0ClientInfo._(JSObject _) implements JSObject {
+  external JSObject?  env;
+  external String  name;
```
The `get` is gone here?

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -96,7 +97,7 @@ class Auth0ClientOptions {
 @JS()
 @anonymous
-class GetTokenSilentlyAuthParams {
+extension type GetTokenSilentlyAuthParams._(JSObject _ )implements JSObject {
```
Missing space after the )

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -165,24 +166,25 @@ class PopupLoginOptions {
-  external dynamic get popup;
-  external int? get timeoutInSeconds;
+extension type PopupConfigOptions._ (JSObject _ ) implements JSObject  {
+  external JSAny  popup;
```
This should be `JSAny?` and there is a space too many. The `get` is also gone here?

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -165,24 +166,25 @@ class PopupLoginOptions {
 
   external factory PopupConfigOptions(
-      {final dynamic popup, final int? timeoutInSeconds});
+      {final JSAny popup, final int? timeoutInSeconds});
```
Shouldn't this be `JSAny?` for the popup argument?

navaronbracke reviewed (changes requested):
Just a quick review, as I have experience with WASM interop from other projects.
(and this might also affect one of my changes to this plugin, hence the added interest)

pmathew92 commented:
> Just a quick review, as I have experience with WASM interop from other projects. (and this might also affect one of my changes to this plugin, hence the added interest)

Hi @navaronbracke  Thank you for reviewing the PR  .Appreciate you taking your time for the same

pmathew92 commented on auth0_flutter/lib/src/web/extensions/web_exception_extensions.dart:
```diff
@@ -1,15 +1,20 @@
 import 'package:auth0_flutter_platform_interface/auth0_flutter_platform_interface.dart';
 
+@JS('Object.keys')
+external JSArray<JSString> keys(final JSObject o);
```
I did initially add this in the js_interop class but the code never got executed when called from other files. Hence moved to respective files. Agreed this needs to be addressed

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -52,29 +52,29 @@ class RedirectLoginOptions {
-  void remove(final String key);
-  void set<T>(final String key, final T entry);
-  Future<List<String>> allKeys();
+extension type  Cache._(JSObject _) implements JSObject {
```
There is an extra space here, after the `extension type ` declaration

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -52,29 +52,29 @@ class RedirectLoginOptions {
-  external String get name;
+extension type Auth0ClientInfo._(JSObject _) implements JSObject {
+  external JSObject? get env;
+  external String  get name;
```
There is an extra space here

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -148,15 +148,15 @@ class LogoutParams {
 @JS()
 @anonymous
-class LogoutOptions {
+extension type  LogoutOptions._(JSObject _) implements JSObject {
```
There is an extra space here, after the `extension type ` declaration

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -106,7 +106,7 @@ class GetTokenSilentlyAuthParams {
+extension type GetTokenSilentlyOptions._(JSObject _) implements JSObject {
   external GetTokenSilentlyAuthParams? get authorizationParams;
   external String? get cacheMode;
   external num? get timeoutInSeconds;
```
This should also be a `JSNumber?` ? Maybe we should do a pass for the remaining `num` declarations.

navaronbracke commented on auth0_flutter/lib/src/web/extensions/web_exception_extensions.dart:
```diff
@@ -1,15 +1,20 @@
-    objectKeys(jsException).forEach((final key) {
-      if (key == 'error' || key == 'error_description') return;
-      details[key as String] = getProperty<dynamic>(jsException, key);
+    keys(jsException.toJSBox).toDart.forEach((final key) {
```
Small suggestion to make this easier to read. We should also not mix up the JS types with the Dart types.
Each Dart string should be manually converted to a JSString. While a JSString _might_ be equal to a String (it is one of the allowed primitives), this might not always be the case and could be JS backend dependent in the future.

```suggestion
    keys(jsException.toJSBox).toDart.forEach((final JSString key) {
       if (key == 'error'.toJS || key == 'error_description'.toJS) return;
       details[key.toDart] = jsException.toJSBox.getProperty<JSAny?>(key);
    });
```

navaronbracke reviewed (changes requested):
Some minor follow-up comments, but this is looking good!

pmathew92 commented:
Hi @navaronbracke  , Would you know what would be the best way to modify the test class 
https://github.com/auth0/auth0-flutter/blob/main/auth0_flutter/test/web/auth0_flutter_web_test.mocks.dart
Currently I am getting the error 
`Classes and mixins can only implement other classes and mixins.`
```class _FakeAuth0Client_0 extends _i1.SmartFake implements _i2.Auth0Client {
  _FakeAuth0Client_0(
    Object parent,
    Invocation parentInvocation,
  ) : super(
          parent,
          parentInvocation,
        );
}```
in the above class declaration as Auth0Client is now an extension type

navaronbracke commented:
Extension type mocking for the new JS types requires some changes.

See the docs at https://dart.dev/interop/js-interop/mock

mockito 5.4.5 did add support for extension types, though, so you can als look into that.

pmathew92 commented:
Hi @navaronbracke  I have updated the PR and fixed the failing UTs
One query , we have this check in the web.dart file which will load the correct file when compiled to WASM
```export 'web/auth0_flutter_plugin_stub.dart'
    if (dart.library.js_interop) 'web/auth0_flutter_plugin_real.dart';```
 But this will fail for users who don't compile to WASM. Is there any way for this check to consider for non-WASM compiles ?

navaronbracke commented:
`dart:html` does not work in the new style JS interop, which is why the import should be changed to `dart.library.js_interop`. The `dart:js_interop` import is available for both the JS backend & the WASM backend.

Since you updated the minimum Dart & Flutter versions in the pubspec, users that want to compile to JS or WASM, using WASM compatible API's are going to need to migrate to that minimum Flutter version.

So I don't see why this would fail if you are running on Flutter 3.24 and trying to compile to JS. Both JS & WASM should work with the new API's.

Widcket commented on auth0_flutter_platform_interface/pubspec.yaml:
```diff
@@ -5,13 +5,14 @@ version: 1.8.0
-  sdk: ">=2.17.0 <4.0.0"
-  flutter: ">=3.0.0"
+  sdk: ">=3.5.0 <4.0.0"
+  flutter: ">=3.24.0"
```
We will need to call out in the release notes this change.

Widcket replied:
Is `3.24.0` the minimum Flutter version we can support?

Widcket replied:
Same for `auth0_flutter/pubspec.yaml`.

pmathew92 replied:
> Is 3.24.0 the minimum Flutter version we can support?

Yes, this is the minimum version required to run WASM compilation
https://docs.flutter.dev/platform-integration/web/wasm#switch-to-the-latest-version-of-flutter

pmathew92 replied:
> We will need to call out this change in the release notes.
Same for auth0_flutter/pubspec.yaml.

Agree. will add this in the release notes while publishing

Widcket commented on auth0_flutter/test/web/extensions/web_exception_extension_test.dart:
```diff
@@ -1,20 +1,19 @@
     if (additionalProps != null) {
       for (final element in additionalProps.entries) {
-        setProperty(jsObject, element.key, element.value);
+        jsObject.setProperty(element.key.toJS, element.value);
```
Should we call `element.value.toJS` here, instead of the the call site in the tests?

pmathew92 replied:
Updated
